lib/imgbundler/imgbundler.go

```diff
@@ -0,0 +1,136 @@
+package imgbundler
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"regexp"
+	"strings"
+	"sync"
+	"time"
+
+	"oss.terrastruct.com/d2/lib/xmain"
+)
+
+var imageRe = regexp.MustCompile(`<image href="([^"]+)"`)
+
+type resp struct {
+	srctxt string
+	data   string
+	err    error
+}
+
+func InlineLocal(ms *xmain.State, in []byte) ([]byte, error) {
+	return inline(ms, in, false)
+}
+
+func InlineRemote(ms *xmain.State, in []byte) ([]byte, error) {
+	return inline(ms, in, true)
+}
+
+func inline(ms *xmain.State, in []byte, isRemote bool) ([]byte, error) {
+	svg := string(in)
+
+	imgs := imageRe.FindAllStringSubmatch(svg, -1)
+
+	var filtered [][]string
+	for _, img := range imgs {
+		u, err := url.Parse(img[1])
+		isRemoteImg := err == nil && strings.HasPrefix(u.Scheme, "http")
+		if isRemoteImg == isRemote {
+			filtered = append(filtered, img)
+		}
+	}
+
+	var wg sync.WaitGroup
+	respChan := make(chan resp)
+
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+	wg.Add(len(filtered))
+	for _, img := range filtered {
+		if isRemote {
+			go fetch(ctx, img[0], img[1], respChan)
+		} else {
+			go read(ctx, img[0], img[1], respChan)
+		}
+	}
+
+	go func() {
+		for {
+			select {
+			case resp, ok := <-respChan:
+				if !ok {
+					return
+				}
+				if resp.err != nil {
+					ms.Log.Error.Printf("image failed to fetch: %s", resp.err.Error())
+				} else {
+					svg = strings.Replace(svg, resp.srctxt, fmt.Sprintf(`<image href="%s"`, resp.data), 1)
+				}
+				wg.Done()
+			}
+		}
+	}()
+
+	wg.Wait()
```

Comment:
I don't like this structure. If fetch/read panic you'll end up with this goroutine freezing forever. Likewise with the above respChan listener one.

You need a semaphore here to limit the number of concurrency requests to something reasonable like 16.

Use `defer` on wg.Done **always** without exception. And do it in the actual goroutine itself, never in a secondary one.

Then what you want to do is run the `wg.Wait` in a separate goroutine and have it close the `respChan` channel when all worker goroutines are done.

This does two things. First, it enables you to correctly return the last non nil error instead of just logging it. Or you could use that multierr package, I forget which one we use in the monorepo, use the same one.
Second, this means you correctly return when all responses are finished processing, not when all worker goroutines are finished which is what's happening now.

Reply:
Nvm, I guess even now you're returning when all worker goroutines are finished but once you move the `wg.Done` into the worker goroutine then you'll definitely need to foreground the respChan goroutine.

Reply:
> You need a semaphore here to limit the number of concurrent requests to something reasonable like 16.

is this just to avoid hogging the user machine resources or is there another reason?

Reply:
Yea, if they have 100 images, you don't want to fire off 100 requests concurrently. It'll saturate their network connection and make all the requests very slow. I know it's counterinuitive as you'd figure things would work faster if you did it with maximum concurrency but things get bottlenecked and it ends up being slower than limiting and using a semaphore.

Reply:
Oh and the semaphore here sould be implemented as a `make(chan struct{}, X)` where X is the number of concurrent requests you want to allow. Then to acquire you'd try to send on the channel (remember to listen on ctx.Done too). To release just read from the channel.

Reply:
> First, it enables you to correctly return the last non nil error instead of just logging it.

the logging is intentional. if a user has 20 icons and one breaks, we tell them but render the rest of the diagram

Reply:
I don't think that should be default behaviour. That should be under a flag to ignore warnings. I don't want to sift through logs to see if my diagram has rendered correctly. I need to be able to rely on the exit status of the CLI.

Reply:
i think this goes back to "warnings > errors" principle. what if they just want to rapidly iterate and start with some paths that they'll then populate with images?

Reply:
We've decided warnings or errors on a case by case basis, not as a general principle. I can think of multiple instances where I've requested errors be returned where there were warnings. And I don't think this is the time for a warning, at least not by default. It's much too easy to miss.

They can leave those paths in comments or use a dummy image URL. Or we can handle `--watch` differently and still render as we can't exit with a status code there anyway.

Reply:
it's not easy to miss at all.

![yo](https://user-images.githubusercontent.com/3120367/204617777-386b2199-9b86-48f8-828d-8c01f4dacee5.png)

even if u make the argument that they're not looking at the render for some reason, the logs are also not easy to miss

<img width="571" alt="Screen Shot 2022-11-29 at 10 41 31 AM" src="https://user-images.githubusercontent.com/3120367/204617937-2fef7c04-cec0-4cc0-8e44-54559429764c.png">

Reply:
u can argue for a lot of unnecessary things if your demographic is unconscious bodies.

and no i think that compromise is even worse. an error should not bail out saving the errored version

can u think of software that produces assets that refuses to work given broken linked assets? i strongly suspect this is not behavior users will desire in practice.

Reply:
> u can argue for a lot of unnecessary things if your demographic is unconscious bodies.

I clearly meant not fully conscious. Not unconscious. Or frankly even newbs.

For example Go doesn't by default just give you a warning if you have an unused variable/import, it errors. Now Go isn't the best example as you can't disable these often annoying ass warnings, but I still strongly prefer it being the default behaviour.

> and no i think that compromise is even worse. an error should not bail out saving the errored version

Ok well what about the `--watch` exception? I think that's perfect.

> can u think of software that produces assets that refuses to work given broken linked assets? i strongly suspect this is not behavior users will desire in practice.

Software package managers. even webpack i think? At least if you're bundling the asset in. Yea so that's the primary difference. We're bundling the asset here, we're not just adding a link.

Oh that reminds me we wanted a `-bundle` flag to bundle in remote links for svgs too right?

Reply:
> yes obviously i did too lol

My man you literally used the word "unconscious body".

> i can see the argument from this angle. i wonder if we should make a --prod flag instead though, because when they're watching with png output, they're not necessarily trying to get production build every time. such a flag would minify svg, error on any broken assets, perhaps do other things.

No, the default should be prod mode. Dev should be the exception. When you compile d2 without any flags, you should know you're getting the safest sanest configuration.

We could add a `--dev` flag but we already have `--watch`. And that's why I'm suggesting reusing `--watch` for now, as that's the only time you'll actually want to ignore such errors and continue rendering.

Reply:
okay. that's fair. doesn't error on watch. errors without. yes?

Reply:
Yup.

Reply:
I find it a bit awkward that an offline tool fails, when it could succeed, if there's no internet connection.
I'd say that we're more into draft/print modes than anything else. While working on my diagram, I might just want to draft it and even replace images with just boxes (to avoid rendering delays, or even downloading the same image multiple times) and get something quickly..

@nhooyr, even using your presentation example. I would be pissed off if my deck software failed to help me build a presentation because it couldn't get an image. Even more if the image source is offline (like the website is down, or some other intermediate network issue). Like, latex will render the PDF and just show a broken image thumbnail. 

My 2 cents here is for "make it work and if you need images strictly, it must be flagged"

Reply:
we're not in disagreement @ejulio-ts 

I want the exit code to be non zero if images fail to be fetched but I'm ok with writing partial output.

Reply:
ok, maybe I missed something on the long read 😂

Reply:
haha no worries, it was my suggested compromise above. But I conceded to Alex as I thought he felt strongly it shouldn't be so based on this response:

> and no i think that compromise is even worse. an error should not bail out saving the errored version